### ISSUE TITLE
Ch3 - questioning readability of point-free style

### ISSUE THREAD

**Issue by mendaomn** [getify/functional-light-js#11]
Hello,
I'm following along, and, as I look at this snippet: 

```
var printIf = reverseArgs(
    uncurry( partialRight( when, output ) )
);
```

I start to wonder whether FP helps or not in writing more readable code. Is it just me or that's a hard to follow line? Is it just a matter of getting used to FP style and will sound obvious to me as my FP skills develop? That looks to me like something I would need to explain to some fellow team member, if I'll ever write anything like it!

What are your thoughts on this? Honestly, that piece of code alone is frightening me a bit, but maybe that's already one of those cases where you suggest that we stop and reconsider point(-ful?) style

**Comment by getify** [comment]
That statement is not terribly self-obvious to me. But I think it *is* more obvious to someone who gets familiar with the different FP operations because each of them jumps out and explains what it's doing.

I think of it sorta like how once you know basic arithmetic, this line makes sense:

```
1 - ((2 * 3^2) / 4)
```

If you don't know arithmetic, that line looks like greek to you, but once you know those operators, it totally falls out how to calculate what it does.

**Comment by mendaomn** [comment]
Got it! 

I think you're right, however this might result in writing some code that will be hard to read and and modify by some other person who's not so familiar with those concepts, I believe.

Is this an actual risk when embracing FP style in your opinion?

Anyway, thank you for writing this book in the open, I am learning a lot and making sense of lots of scattered notions I have on the topic. 

As a final question, is there in your github a module/piece of software that you developed entirely or partly in FP style? I would love to see what for form it takes in a real project :)

Thank you for sharing and taking the time to answer this, keep up!

**Comment by yurrriq** [comment]
As an FPer and not a JSer, and without context, the listing above means almost nothing to me. I don't know what `partialRight` is, for example. I'm sure it makes sense in the text.

In general though, I think point-free style can serve to add focus to *verbs* (functions) rather than *nouns* (values) which can often add clarity. What's more, a lot of the time, the "point-freeing" is simply η-reduction.

Take, for example, a function `f` that takes a list of numbers and returns the list of their absolute values:

```haskell
f : Num a => [a] -> [a]
f xs = map (\x -> abs x) xs
```

In point-free style, `f` could be rewritten:

```haskell
f = map abs
```

In my opinion, the point-free version of `f` communicates its meaning more clearly.

It's easy to take it too far and render your code "pointless," but I think point-free style is a valuable tool too have and use in moderation.

**Comment by getify** [comment]
I would welcome any feedback or suggestions on ways to improve the illustration of point-free style in this case. Of course, [the whole section of the text](https://github.com/getify/functional-light-js/blob/master/ch3.md#no-points) should first be read to understand the context it comes from.

Briefly, this function is presented:

```js
function printIf( msg, predicate ) {
    if (predicate( msg )) {
        output( msg );
    }
}
```

It is observed that this function (both its behavior and its signature) can be expressed as a combination of other well-known FP operations, thereby achieving a point-free representation of it.

Earlier in the text, `reverseArgs(..)` (which I believe is maybe commonly called `flip(..)` in the FP realm) is defined as:

```js
var reverseArgs = fn => (...args) => fn( ...args.reverse() );
```

And then `partialRight(..)` is defined as a convenience for doing partial-application but from the right instead of the left:

```js
var partialRight = ( fn, ...args ) => reverseArgs( partial( reverseArgs( fn ), ...args ) );
```

The `when(..)` is pretty similar to how it shows up in common FP libs:

```js
var when = (predicate,fn) => (...args) => predicate( ...args ) ? fn( ...args ) : undefined;
```

So, putting these altogether, we get:

```js
var printIf = reverseArgs( uncurry( partialRight( when, output ) ) );
```

Or to use more common FP utility names and not use the convenience `partialRight(..)`, we could start with:

```js
var printIf = partial( flip( when ), output );

printIf( predicateFn )( str );
```

But that signature doesn't match the original, so first `uncurry(..)` then `flip(..)`:

```js
var printIf = flip( uncurry( partial( flip( when ), output ) ) );

// compared to:

var printIf = reverseArgs( uncurry( partialRight( when, output ) ) );
```

**Comment by getify** [comment]
> I think point-free style is a valuable tool to have and use in moderation.

> maybe that's already one of those cases where you suggest that we stop and reconsider point(-ful?) style

The text does make this admonition right off the bat:

> **Warning:** Stop for a moment. Let's make sure we're careful not to take this discussion as an unbounded suggestion that you go overboard trying to be point-free in your FP code at all costs. This should be a technique for improving readability, when used in moderation. But as with most things in software development, you can definitely abuse it. If your code gets harder to understand because of the hoops you have to jump through to be point-free, stop. You won't win a blue ribbon just because you found some clever but esoteric way to remove another "point" from your code.

So, clearly I agree there is need for moderation. I personally think the specific `printIf(..)` example presented is still in the neighborhood of useful/salient, but I recognize these are judgement calls, and I invite the reader to make their own determinations.

**Comment by getify** [comment]
> is there in your github a module/piece of software that you developed entirely or partly in FP style?

I don't have any such thing yet. But it's a good idea that I will hopefully fulfill at some point soon. :)

**Comment by yurrriq** [comment]
@getify, thanks for that explanation. I think it was too late at night for me when I made my comment. The `reverseArgs`/`flip` connection makes perfect sense and now I can read the original listing.

I should really read the text before commenting again, but I've been enjoying the discussions in issues. :smile: 

Keep up the good work!

**Comment by getify** [comment]
Thanks for the discussion here! :)
